codecov_auth/tests/unit/test_repo_authentication.py

```diff
@@ -341,3 +345,226 @@ def test_orgleveltoken_success_auth_enterprise(self, db, mocker):
         assert auth.allows_repo(repository)
         assert auth.allows_repo(other_repo_from_owner)
         assert not auth.allows_repo(random_repo)
+
+
+class TestTokenlessAuth(object):
+    @pytest.mark.parametrize(
+        "headers",
+        [{}, {"X-Tokenless": "user-name/repo-forked"}, {"X-Tokenless-PR": "15"}],
+    )
+    def test_tokenless_missing_headers(self, headers):
+        request = APIRequestFactory().post(
+            "/upload/unknown_provider/owner::::repo/commits/commit_sha/reports/report_code/uploads",
+            headers=headers,
+        )
+        authentication = TokenlessAuthentication()
+        res = authentication.authenticate(request)
+        assert res is None
+
+    def test_tokenless_bad_path(self):
+        request = APIRequestFactory().post(
+            "/endpoint",
+            headers={"X-Tokenless": "user-name/repo-forked", "X-Tokenless-PR": "15"},
+        )
+        authentication = TokenlessAuthentication()
+        with pytest.raises(exceptions.AuthenticationFailed) as exp:
+            authentication.authenticate(request)
+        assert str(exp.value) == "Not valid tokenless upload"
+
+    def test_tokenless_unknown_service(self):
+        request = APIRequestFactory().post(
+            "/upload/unknown_provider/owner::::repo/commits/commit_sha/reports/report_code/uploads",
+            headers={"X-Tokenless": "user-name/repo-forked", "X-Tokenless-PR": "15"},
+        )
+        authentication = TokenlessAuthentication()
+        with pytest.raises(exceptions.AuthenticationFailed) as exp:
+            authentication.authenticate(request)
+        assert str(exp.value) == "Not valid tokenless upload"
+
+    def test_tokenless_not_supported_services(self):
```

Comment:
Does tokenless support `gh` and `github` and all there uppercase/lowercase variants?

Reply:
depends if the upload URL supports that or not. I don't think it does.
If you look in the `GetterMixin` ([here](https://github.com/codecov/codecov-api/blob/c69525cafd7d744437798990b845b10b2f19d74b/upload/views/base.py#L31)) you'll see that the `Service` has a limited number of options.

Reply:
Good to know in case we ever want to support/change for multiple services